DocSearch.ipynb

```diff
@@ -0,0 +1,145 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook is a demonstration for how to use ThirdAI's document search engine. It assumes you are running in the ThirdAI demo docker container. Once in the container, you can get this notebook by running\n",
+    "```\n",
+    "wget https://raw.githubusercontent.com/ThirdAILabs/Demos/main/DocSearch.ipynb\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we're going to import our document search package and the embedding model we will use to embed the documents, and a couple of other things that will help us along the way."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "\n",
+    "import warnings\n",
+    "warnings.filterwarnings('ignore')\n",
+    "\n",
+    "import numpy as np\n",
+    "\n",
+    "import thirdai\n",
+    "from thirdai.search import DocRetrieval\n",
+    "\n",
+    "from embeddings import DocSearchModel\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    " Now you're going to need a dataset, which just consists of a collection of (document_id, document_text) pairs. Our engine works best with document sizes between 20 and 200 words long. Feel free to split big documents up into many small passages!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# TODO for you: Populate dataset (choose whatever you want!)\n",
+    "dataset = []"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's create an initially empty index! As a quick description of the input to the constructor, \n",
+    "* dense_input_dimension is the dimension of the output of our embedding model (128 with our preloaded model)\n",
+    "\n",
+    "* num_tables and hashes_per_table are hyperparameters of the index\n",
+    "  * increasing num_tables increases the accuracy of the model at the cost of speed and memory (a good value is the prepopulated 16)\n",
+    "  \n",
+    "  * hashes_per_table has a sweet spot in accuracy at around log_2(average document size)\n",
+    "  \n",
+    "* centroids are a numpy array that represents precomputed centroids for the embedding space; we've already calculated and stored these for you! We've precomputed 2^18 of them, but if you're only adding a few points a simple heuristic is to select a random subsample of them, which we do below."
```

Comment:
I thought we have an api that doesn't take in centroids?

Reply:
Nope, not yet (but we pregenerate centroids for people)